From 5220454701112b2fb39bb81c5b2de443174653cd Mon Sep 17 00:00:00 2001
From: Yann Gautier <yann.gautier@foss.st.com>
Date: Tue, 12 Sep 2023 17:40:32 +0200
Subject: [PATCH 412/841] feat(st-clock): stub some registers access when not
 TDCID

If Cortex-A35 is not TDCID, several registers (or register fields)
in RCC are no more accessible.
Stub their access under STM32MP_M33_TDCID flag.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 83aa30a0c43127a2da2cec5e04ae1f6957da6873 feat(st-clock): stub some registers access when not TDCID]
Signed-off-by: Yann Gautier <yann.gautier@foss.st.com>
Change-Id: I1a9e724ebe5d10de53224c348a755be3bb6f4bdf
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/328517
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Gabriel FERNANDEZ <gabriel.fernandez@foss.st.com>
Reviewed-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
---
 drivers/st/clk/stm32mp2_clk.c | 27 ++++++++++++++++++++++++++-
 1 file changed, 26 insertions(+), 1 deletion(-)

diff --git a/drivers/st/clk/stm32mp2_clk.c b/drivers/st/clk/stm32mp2_clk.c
index 4df5051de..01eb08919 100644
--- a/drivers/st/clk/stm32mp2_clk.c
+++ b/drivers/st/clk/stm32mp2_clk.c
@@ -964,6 +964,7 @@ static const struct stm32_clk_ops clk_stm32_flexgen_ops = {
 #define RCC_16_MHZ	UL(16000000)
 
 #ifdef IMAGE_BL2
+#if !STM32MP_M33_TDCID
 static int clk_stm32_osc_msi_set_rate(struct stm32_clk_priv *priv, int id, unsigned long rate,
 				      unsigned long prate)
 {
@@ -988,6 +989,7 @@ static int clk_stm32_osc_msi_set_rate(struct stm32_clk_priv *priv, int id, unsig
 
 	return ret;
 }
+#endif
 #endif /* IMAGE_BL2 */
 
 static unsigned long clk_stm32_osc_msi_recalc_rate(struct stm32_clk_priv *priv,
@@ -1006,8 +1008,10 @@ static unsigned long clk_stm32_osc_msi_recalc_rate(struct stm32_clk_priv *priv,
 static const struct stm32_clk_ops clk_stm32_osc_msi_ops = {
 	.recalc_rate	= clk_stm32_osc_msi_recalc_rate,
 	.is_enabled	= clk_stm32_osc_gate_is_enabled,
+#if !STM32MP_M33_TDCID
 	.enable		= clk_stm32_osc_gate_enable,
 	.disable	= clk_stm32_osc_gate_disable,
+#endif
 	.init		= clk_stm32_osc_init,
 };
 
@@ -1352,7 +1356,7 @@ static void stm32mp2_a35_ss_on_hsi(void)
 #ifdef IMAGE_BL2
 static void stm32mp2_clk_muxsel_on_hsi(struct stm32_clk_priv *priv)
 {
-
+#if !STM32MP_M33_TDCID
 	mmio_clrbits_32(priv->base + RCC_MUXSELCFGR,
 			RCC_MUXSELCFGR_MUXSEL0_MASK |
 			RCC_MUXSELCFGR_MUXSEL1_MASK |
@@ -1362,8 +1366,14 @@ static void stm32mp2_clk_muxsel_on_hsi(struct stm32_clk_priv *priv)
 			RCC_MUXSELCFGR_MUXSEL5_MASK |
 			RCC_MUXSELCFGR_MUXSEL6_MASK |
 			RCC_MUXSELCFGR_MUXSEL7_MASK);
+#else
+	/* MUXSEL5 for PLL1 selection = CA35 clock source, always accessible */
+	mmio_clrbits_32(priv->base + RCC_MUXSELCFGR,
+			RCC_MUXSELCFGR_MUXSEL5_MASK);
+#endif
 }
 
+#if !STM32MP_M33_TDCID
 static void stm32mp2_clk_xbar_on_hsi(struct stm32_clk_priv *priv)
 {
 	uintptr_t xbar0cfgr = priv->base + RCC_XBAR0CFGR;
@@ -1375,6 +1385,7 @@ static void stm32mp2_clk_xbar_on_hsi(struct stm32_clk_priv *priv)
 				   XBAR_SRC_HSI);
 	}
 }
+#endif
 #endif /* IMAGE_BL2 */
 
 /* TODO: MOVE THIS FUNCTION A35 ONLY */
@@ -1672,6 +1683,7 @@ static int clk_stm32_pll_init(struct stm32_clk_priv *priv, int pll_idx)
 
 static int stm32mp2_clk_pll_configure(struct stm32_clk_priv *priv)
 {
+#if !STM32MP_M33_TDCID
 	enum pll_id i;
 	int err;
 
@@ -1683,6 +1695,9 @@ static int stm32mp2_clk_pll_configure(struct stm32_clk_priv *priv)
 	}
 
 	return 0;
+#else
+	return clk_stm32_pll_init(priv, _PLL1);
+#endif
 }
 
 static int wait_predivsr(uint16_t channel)
@@ -1841,6 +1856,7 @@ static int stm32mp2_clk_flexgen_configure(struct stm32_clk_priv *priv)
 	return 0;
 }
 
+#if !STM32MP_M33_TDCID
 static void stm32_enable_oscillator_hse(struct stm32_clk_priv *priv)
 {
 	struct stm32_clk_platdata *pdata = priv->pdata;
@@ -1884,16 +1900,20 @@ static void stm32_enable_oscillator_lse(struct stm32_clk_priv *priv)
 
 	_clk_stm32_gate_enable(priv, osc_data->gate_id);
 }
+#endif
 
 static int stm32mp2_clk_switch_to_hsi(struct stm32_clk_priv *priv)
 {
 	stm32mp2_a35_ss_on_hsi();
 	stm32mp2_clk_muxsel_on_hsi(priv);
+#if !STM32MP_M33_TDCID
 	stm32mp2_clk_xbar_on_hsi(priv);
+#endif
 
 	return 0;
 }
 
+#if !STM32MP_M33_TDCID
 static int stm32_clk_oscillators_wait_lse_ready(struct stm32_clk_priv *priv)
 {
 	int ret = 0;
@@ -1928,6 +1948,7 @@ static void stm32_clk_oscillators_enable(struct stm32_clk_priv *priv)
 	stm32_enable_oscillator_msi(priv);
 	_clk_stm32_enable(priv, _CK_LSI);
 }
+#endif
 
 static int stm32_clk_configure_div(struct stm32_clk_priv *priv, uint32_t data)
 {
@@ -2050,7 +2071,9 @@ static int stm32mp2_init_clock_tree(void)
 	stm32mp_stgen_restore_rate();
 	generic_delay_timer_init();
 
+#if !STM32MP_M33_TDCID
 	stm32_clk_oscillators_enable(priv);
+#endif
 
 	/* Come back to HSI */
 	ret = stm32mp2_clk_switch_to_hsi(priv);
@@ -2063,11 +2086,13 @@ static int stm32mp2_init_clock_tree(void)
 		panic();
 	}
 
+#if !STM32MP_M33_TDCID
 	/* Wait LSE ready before to use it */
 	ret = stm32_clk_oscillators_wait_lse_ready(priv);
 	if (ret != 0) {
 		panic();
 	}
+#endif
 
 	ret = stm32mp2_clk_flexgen_configure(priv);
 	if (ret != 0) {
-- 
2.47.2

